dlt/destinations/path_utils.py

```diff
@@ -1,45 +1,205 @@
+    }
+)
 
-# TODO: ensure layout only has supported placeholders
-SUPPORTED_PLACEHOLDERS = {"schema_name", "table_name", "load_id", "file_id", "ext", "curr_date"}
 
 SUPPORTED_TABLE_NAME_PREFIX_PLACEHOLDERS = ("schema_name",)
 
 
-def check_layout(layout: str) -> List[str]:
+def get_placeholders(layout: str) -> List[str]:
+    return re.findall(r"\{(.*?)\}", layout)
+
+
+def prepare_datetime_params(
+    current_datetime: Optional[pendulum.DateTime] = None,
+    load_package_timestamp: Optional[str] = None,
+) -> Dict[str, str]:
+    params: Dict[str, str] = {}
+    current_timestamp: pendulum.DateTime = None
+    if load_package_timestamp:
+        current_timestamp = pendulum.parse(load_package_timestamp)  # type: ignore[assignment]
+        params["load_package_timestamp"] = str(int(current_timestamp.timestamp()))
+
+    if not current_datetime:
+        if current_timestamp:
+            logger.info("current_datetime is not set, using timestamp from load package")
+            current_datetime = current_timestamp
+        else:
+            logger.info("current_datetime is not set, using pendulum.now()")
+            current_datetime = pendulum.now()
+
+    params["timestamp"] = str(int(current_datetime.timestamp()))
+    params["curr_date"] = str(current_datetime.date())
+
+    for format_string in DATETIME_PLACEHOLDERS:
+        params[format_string] = current_datetime.format(format_string).lower()
+
+    return params
+
+
+def prepare_params(
+    extra_placeholders: Optional[Dict[str, Any]] = None,
+    job_info: Optional[ParsedLoadJobFileName] = None,
+    schema_name: Optional[str] = None,
+    load_id: Optional[str] = None,
+) -> Dict[str, Any]:
+    params: Dict[str, Any] = {}
+    table_name = None
+    file_id = None
+    ext = None
+    if job_info:
+        table_name = job_info.table_name
+        file_id = job_info.file_id
+        ext = job_info.file_format
+        params.update(
+            {
+                "table_name": table_name,
+                "file_id": file_id,
+                "ext": ext,
+            }
+        )
+
+    if schema_name:
+        params["schema_name"] = schema_name
+
+    if load_id:
+        params["load_id"] = load_id
+
+    # Resolve extra placeholders
+    if extra_placeholders:
+        for key, value in extra_placeholders.items():
+            if callable(value):
+                try:
+                    params[key] = value(
+                        schema_name,
+                        table_name,
+                        load_id,
+                        file_id,
+                        ext,
+                    )
+                except TypeError as exc:
+                    raise InvalidPlaceholderCallback(key) from exc
+            else:
+                params[key] = value
+
+    return params
+
+
+def check_layout(
+    layout: str,
+    extra_placeholders: Optional[Dict[str, Any]] = None,
+    allowed_placeholders: Optional[Set[str]] = SUPPORTED_PLACEHOLDERS,
```

Comment:
why allowed placeholders are supported placeholders? I'd rename them to standard placeholders

Reply:
They were named as such in the old code so I left it.
I also agree that standard_placeholders sounds and reads better.